File: 14_2-D_Dynamic_Programming/05_Target_Sum/0494-target-sum.py
"""
Problem: LeetCode 494 - Target Sum

Key Idea:
The key idea is to use dynamic programming to count the number of ways to reach a certain sum by assigning either a positive or negative sign to each element in the input array.

Time Complexity:
The time complexity is O(n * sum), where 'n' is the number of elements in the input array 'nums' and 'sum' is the sum of all elements in 'nums'. We iterate through 'n' elements and, for each element, iterate through 'sum' values.

Space Complexity:
The space complexity is O(sum) because we use a 1D array 'dp' of size 'sum + 1' to store the number of ways to reach each sum from 0 to 'sum'.
"""


class Solution:
    def findTargetSumWays(self, nums: List[int], S: int) -> int:
        # Calculate the sum of all elements in the input array 'nums'.
        total_sum = sum(nums)

        # If the total sum is less than the target sum 'S', it's not possible to reach 'S'.
        if (total_sum + S) % 2 != 0 or total_sum < abs(S):
            return 0

        # Calculate the target sum for positive signs. (total_sum + S) / 2
        target = (total_sum + S) // 2

        # Initialize a 1D array 'dp' to store the number of ways to reach each sum from 0 to 'target'.
        dp = [0] * (target + 1)

        # There is one way to reach a sum of 0 (by not selecting any element).
        dp[0] = 1

        # Iterate through each element in the input array 'nums'.
        for num in nums:
            # Update the 'dp' array for each sum from 'target' to 'num'.
            for i in range(target, num - 1, -1):
                dp[i] += dp[i - num]

        # The 'dp[target]' contains the number of ways to reach the target sum 'S'.
        return dp[target]

File: 14_2-D_Dynamic_Programming/05_Target_Sum/0494-target-sum.rs
/*
Problem: LeetCode 494 - Target Sum

Key Idea:
The key idea is to use dynamic programming to count the number of ways to achieve a target sum using the given positive and negative numbers.

Approach:
1. Initialize a variable `sum` to store the sum of all elements in the input array `nums`.
2. Calculate the target sum `target` as the difference between the `sum` and the given `S`.
3. Check if `target` is odd or negative. If it is, there are no ways to achieve the target sum, so return 0.
4. Initialize a vector `dp` of size `2 * target + 1` to store the number of ways to achieve each possible sum from `-target` to `target`.
5. Initialize `dp[0]` as 1 because there is one way to achieve a sum of 0 (by not choosing any number).
6. Iterate through the elements of `nums`:
   - For each element `num`, iterate through the `dp` vector from `target` down to `num` (inclusive).
   - For each index `i`, update `dp[i]` by adding `dp[i - num]`. This represents the number of ways to achieve the sum `i` by including the current element.
7. After the iteration, `dp[target]` will contain the number of ways to achieve the target sum.
8. Return `dp[target]` as the result.

Time Complexity:
O(n * target), where `n` is the number of elements in the input array `nums` and `target` is the target sum. We fill in the `dp` table with a nested loop.

Space Complexity:
O(target), as we use a vector of size `2 * target + 1` to store the dynamic programming values.
*/

impl Solution {
    pub fn find_target_sum_ways(nums: Vec<i32>, s: i32) -> i32 {
        let sum: i32 = nums.iter().sum();
        let target = sum - s;

        if target < 0 || target % 2 != 0 {
            return 0;
        }

        let target = (target / 2) as usize;
        let mut dp = vec![0; target + 1];
        dp[0] = 1;

        for num in nums {
            for i in (num as usize..=target).rev() {
                dp[i] += dp[i - num as usize];
            }
        }

        dp[target]
    }
}
